File: fe2-android/app/src/test/ui/robolectric/com/github/dedis/popstellar/ui/lao/event/rollcall/RollCallAddAttendeeTest.java
package com.github.dedis.popstellar.ui.lao.event.rollcall;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.filters.LargeTest;

import com.github.dedis.popstellar.model.objects.Lao;
import com.github.dedis.popstellar.model.objects.security.KeyPair;
import com.github.dedis.popstellar.model.objects.security.PublicKey;
import com.github.dedis.popstellar.model.objects.view.LaoView;
import com.github.dedis.popstellar.repository.LAORepository;
import com.github.dedis.popstellar.testutils.Base64DataUtils;
import com.github.dedis.popstellar.testutils.BundleBuilder;
import com.github.dedis.popstellar.testutils.fragment.ActivityFragmentScenarioRule;
import com.github.dedis.popstellar.ui.lao.LaoActivity;
import com.github.dedis.popstellar.ui.qrcode.QrScannerFragment;
import com.github.dedis.popstellar.ui.qrcode.ScanningAction;
import com.github.dedis.popstellar.utility.error.UnknownLaoException;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.ExternalResource;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoTestRule;

import dagger.hilt.android.testing.*;
import io.reactivex.subjects.BehaviorSubject;

import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static com.github.dedis.popstellar.testutils.Base64DataUtils.generateKeyPair;
import static com.github.dedis.popstellar.testutils.UITestUtils.forceTypeText;
import static com.github.dedis.popstellar.testutils.pages.lao.LaoActivityPageObject.containerId;
import static com.github.dedis.popstellar.testutils.pages.lao.LaoActivityPageObject.laoIdExtra;
import static com.github.dedis.popstellar.testutils.pages.scanning.QrScanningPageObject.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.when;

@LargeTest
@HiltAndroidTest
@RunWith(AndroidJUnit4.class)
public class RollCallAddAttendeeTest {

  private static final String LAO_NAME = "lao";
  private static final KeyPair SENDER_KEY = generateKeyPair();
  private static final PublicKey SENDER = SENDER_KEY.publicKey;
  private static final Lao LAO = new Lao(LAO_NAME, SENDER, 10223421);
  private static final String LAO_ID = LAO.getId();
  private static final String POP_TOKEN =
      Base64DataUtils.generatePoPToken().publicKey.getEncoded();
  private static final String VALID_RC_MANUAL_INPUT = "{\"pop_token\": \"" + POP_TOKEN + "\"}";
  public static final String JSON_INVALID_INPUT = "{pop_token:" + POP_TOKEN;
  public static final String VALID_WITNESS_MANUAL_INPUT =
      "{\"main_public_key\": \"" + POP_TOKEN + "\"}";
  public static final String INVALID_KEY_FORMAT_INPUT = "{\"pop_token\": \"invalid_key\"}";

  private static final BehaviorSubject<LaoView> laoSubject =
      BehaviorSubject.createDefault(new LaoView(LAO));

  @BindValue @Mock LAORepository repository;

  @Rule public InstantTaskExecutorRule rule = new InstantTaskExecutorRule();

  @Rule(order = 0)
  public final MockitoTestRule mockitoRule = MockitoJUnit.testRule(this);

  @Rule(order = 1)
  public final HiltAndroidRule hiltRule = new HiltAndroidRule(this);

  @Rule(order = 2)
  public final ExternalResource setupRule =
      new ExternalResource() {
        @Override
        protected void before() throws UnknownLaoException {
          hiltRule.inject();
          when(repository.getLaoObservable(anyString())).thenReturn(laoSubject);
          when(repository.getLaoView(any())).thenReturn(new LaoView(LAO));
        }
      };

  @Rule(order = 3)
  public ActivityFragmentScenarioRule<LaoActivity, QrScannerFragment> activityScenarioRule =
      ActivityFragmentScenarioRule.launchIn(
          LaoActivity.class,
          new BundleBuilder().putString(laoIdExtra(), LAO_ID).build(),
          containerId(),
          QrScannerFragment.class,
          () -> QrScannerFragment.newInstance(ScanningAction.ADD_ROLL_CALL_ATTENDEE));

  @Test
  public void addButtonIsDisplayed() {
    openManualButton().check(matches(isDisplayed()));
  }

  @Test
  public void addingAttendeeManuallyUpdatesCount() {
    openManualButton().perform(click());
    manualAddEditText().perform(forceTypeText(VALID_RC_MANUAL_INPUT));
    manualAddConfirm().perform(click());

    // Since we haven't mocked for the viewModel to fetch the organizer token, adding an attendee
    // should result in a total of one attendee
    attendeeCount().check(matches(withText("1")));
  }

  @Test
  public void addingInvalidJsonFormatDoesNotAddAttendees() {
    openManualButton().perform(click());
    manualAddEditText().perform(forceTypeText(JSON_INVALID_INPUT));
    manualAddConfirm().perform(click());

    attendeeCount().check(matches(withText("0")));
  }

  @Test
  public void addingValidNonRcFormatDoesNotAddAttendees() {
    openManualButton().perform(click());
    manualAddEditText().perform(forceTypeText(VALID_WITNESS_MANUAL_INPUT));
    manualAddConfirm().perform(click());

    attendeeCount().check(matches(withText("0")));
  }

  @Test
  public void addingKeyFormatDoesNotAddAttendees() {
    openManualButton().perform(click());
    manualAddEditText().perform(forceTypeText(INVALID_KEY_FORMAT_INPUT));
    manualAddConfirm().perform(click());

    attendeeCount().check(matches(withText("0")));
  }
}

File: fe2-android/app/src/test/ui/robolectric/com/github/dedis/popstellar/ui/lao/event/rollcall/RollCallAddAttendeeTest.kt
package com.github.dedis.popstellar.ui.lao.event.rollcall

import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions
import androidx.test.espresso.matcher.ViewMatchers
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.LargeTest
import com.github.dedis.popstellar.R
import com.github.dedis.popstellar.model.objects.Lao
import com.github.dedis.popstellar.model.objects.view.LaoView
import com.github.dedis.popstellar.repository.LAORepository
import com.github.dedis.popstellar.testutils.Base64DataUtils
import com.github.dedis.popstellar.testutils.BundleBuilder
import com.github.dedis.popstellar.testutils.MockitoKotlinHelpers
import com.github.dedis.popstellar.testutils.UITestUtils
import com.github.dedis.popstellar.testutils.UITestUtils.forceTypeText
import com.github.dedis.popstellar.testutils.fragment.ActivityFragmentScenarioRule
import com.github.dedis.popstellar.testutils.pages.lao.LaoActivityPageObject
import com.github.dedis.popstellar.testutils.pages.scanning.QrScanningPageObject
import com.github.dedis.popstellar.ui.lao.LaoActivity
import com.github.dedis.popstellar.ui.qrcode.QrScannerFragment
import com.github.dedis.popstellar.ui.qrcode.QrScannerFragment.Companion.newInstance
import com.github.dedis.popstellar.ui.qrcode.ScanningAction
import com.github.dedis.popstellar.utility.error.UnknownLaoException
import dagger.hilt.android.testing.BindValue
import dagger.hilt.android.testing.HiltAndroidRule
import dagger.hilt.android.testing.HiltAndroidTest
import io.reactivex.subjects.BehaviorSubject
import org.junit.Assert
import org.junit.Ignore
import org.junit.Rule
import org.junit.Test
import org.junit.rules.ExternalResource
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnit
import org.mockito.junit.MockitoTestRule

@LargeTest
@HiltAndroidTest
@RunWith(AndroidJUnit4::class)
class RollCallAddAttendeeTest {
  @BindValue @Mock lateinit var repository: LAORepository

  @JvmField @Rule var rule = InstantTaskExecutorRule()

  @JvmField @Rule(order = 0) val mockitoRule: MockitoTestRule = MockitoJUnit.testRule(this)

  @JvmField @Rule(order = 1) val hiltRule = HiltAndroidRule(this)

  @JvmField
  @Rule(order = 2)
  val setupRule: ExternalResource =
    object : ExternalResource() {
      @Throws(UnknownLaoException::class)
      override fun before() {
        hiltRule.inject()
        Mockito.`when`(repository.getLaoObservable(ArgumentMatchers.anyString()))
          .thenReturn(laoSubject)
        Mockito.`when`(repository.getLaoView(MockitoKotlinHelpers.any())).thenReturn(LaoView(LAO))
      }
    }

  @JvmField
  @Rule(order = 3)
  var activityScenarioRule: ActivityFragmentScenarioRule<LaoActivity, QrScannerFragment> =
    ActivityFragmentScenarioRule.launchIn(
      LaoActivity::class.java,
      BundleBuilder().putString(LaoActivityPageObject.laoIdExtra(), LAO_ID).build(),
      LaoActivityPageObject.containerId(),
      QrScannerFragment::class.java
    ) {
      newInstance(ScanningAction.ADD_ROLL_CALL_ATTENDEE)
    }

  @Test
  fun addButtonIsDisplayed() {
    QrScanningPageObject.openManualButton()
      .check(ViewAssertions.matches(ViewMatchers.isDisplayed()))
  }

  @Test
  fun addingAttendeeManuallyUpdatesCount() {
    QrScanningPageObject.openManualButton().perform(ViewActions.click())

    val input = QrScanningPageObject.manualInputWithHintRes(R.string.manual_add_rc_add_hint)
    Assert.assertNotNull(input)
    input.perform(forceTypeText(VALID_RC_MANUAL_INPUT))

    QrScanningPageObject.manualAddConfirm().perform(ViewActions.click())

    // Since we haven't mocked for the viewModel to fetch the organizer token, adding an attendee
    // should result in a total of one attendee
    QrScanningPageObject.attendeeCount().check(ViewAssertions.matches(ViewMatchers.withText("1")))
  }

  @Test
  fun addingSameAttendeeTwiceDoesNotIncreaseCount() {
    QrScanningPageObject.openManualButton().perform(ViewActions.click())
    val input = QrScanningPageObject.manualInputWithHintRes(R.string.manual_add_rc_add_hint)
    Assert.assertNotNull(input)
    input.perform(forceTypeText(VALID_RC_MANUAL_INPUT))
    QrScanningPageObject.manualAddConfirm().perform(ViewActions.click())
    QrScanningPageObject.attendeeCount().check(ViewAssertions.matches(ViewMatchers.withText("1")))

    QrScanningPageObject.manualAddConfirm().perform(ViewActions.click())
    QrScanningPageObject.attendeeCount().check(ViewAssertions.matches(ViewMatchers.withText("1")))
    UITestUtils.assertToastIsDisplayedWithText(R.string.attendee_already_scanned_warning)
  }

  @Test
  fun addingInvalidKeyDoesNotAddAttendees() {
    QrScanningPageObject.openManualButton().perform(ViewActions.click())
    val input = QrScanningPageObject.manualInputWithHintRes(R.string.manual_add_rc_add_hint)
    Assert.assertNotNull(input)
    input.perform(forceTypeText(INVALID_KEY_RC_MANUAL_INPUT))
    QrScanningPageObject.manualAddConfirm().perform(ViewActions.click())
    QrScanningPageObject.attendeeCount().check(ViewAssertions.matches(ViewMatchers.withText("0")))
    UITestUtils.assertToastIsDisplayedWithText(R.string.qr_code_not_pop_token)
  }

  @Ignore("This test is failing because any public key looking input is accepted... Thus LAOIDs are accepted as well, but should not. Also, Attendees that are not in the LAO are accepted as well but should not.")
  @Test
  fun addingValidNonRcFormatDoesNotAddAttendees() {
    QrScanningPageObject.openManualButton().perform(ViewActions.click())
    val input = QrScanningPageObject.manualInputWithHintRes(R.string.manual_add_rc_add_hint)
    Assert.assertNotNull(input)
    input.perform(forceTypeText(VALID_LAO_ID_INPUT))
    QrScanningPageObject.manualAddConfirm().perform(ViewActions.click())
    QrScanningPageObject.attendeeCount().check(ViewAssertions.matches(ViewMatchers.withText("0")))
    UITestUtils.assertToastIsDisplayedWithText(R.string.qr_code_not_pop_token)
  }

  @Test
  fun addingEmptyKeyDoesNotAddAttendees() {
    QrScanningPageObject.openManualButton().perform(ViewActions.click())
    val input = QrScanningPageObject.manualInputWithHintRes(R.string.manual_add_rc_add_hint)
    Assert.assertNotNull(input)
    input.perform(forceTypeText(EMPTY_KEY_INPUT))
    QrScanningPageObject.manualAddConfirm().perform(ViewActions.click())
    QrScanningPageObject.attendeeCount().check(ViewAssertions.matches(ViewMatchers.withText("0")))
    UITestUtils.assertToastIsDisplayedWithText(R.string.qrcode_scanning_manual_entry_error)
  }

  companion object {
    private const val LAO_NAME = "lao"
    private val SENDER_KEY = Base64DataUtils.generateKeyPair()
    private val SENDER = SENDER_KEY.publicKey
    private val LAO = Lao(LAO_NAME, SENDER, 10223421)
    private val LAO_ID = LAO.id
    private val POP_TOKEN = Base64DataUtils.generatePoPToken().publicKey.encoded
    private val VALID_RC_MANUAL_INPUT = POP_TOKEN
    val INVALID_KEY_RC_MANUAL_INPUT = "invalid for sure"
    val VALID_LAO_ID_INPUT = LAO_ID
    const val EMPTY_KEY_INPUT = ""
    private val laoSubject = BehaviorSubject.createDefault(LaoView(LAO))
  }
}
